### ISSUE TITLE
ko v0.12.0 and AWS ECR fails at SBOM upload with: Invalid JSON syntax

### ISSUE THREAD

**Issue by nairb774** [ko-build/ko#878]
I'm running into an issue using ko v0.12.0, go 1.18.7, and AWS's ECR. ECR is generating the following error output:

> Using base 123456789012.dkr.ecr.us-east-1.amazonaws.com/mirror/gcr.io/distroless/static-debian10:nonroot@sha256:df44f5173aa356eb3162c58cb958181cebca72ae55904cd963c1773e6da14a25 for github.com/some/path
> Building github.com/some/path for linux/amd64
> Publishing 123456789012.dkr.ecr.us-east-1.amazonaws.com/github.com/some/path:1.4.3
> pushed blob: sha256:e5b55a9d6a022a9d29060dd1ed44f416838d064208672cd5bf73ad37ace75d08
> pushed blob: sha256:5f15857050e2b1cc6ab67e71d4945f16f7553044163c0caf8a6afb0157ca536e
> Error: failed to publish images: error publishing ko://github.com/some/path: writing sbom: PUT https://123456789012.dkr.ecr.us-east-1.amazonaws.com/v2/github.com/some/path/manifests/sha256-a889960bc4557663c9541f6876f3a701708c7f886757a7a407537f8625b7ea9b.sbom: UNSUPPORTED: Invalid parameter at 'ImageManifest' failed to satisfy constraint: 'Invalid JSON syntax'
> error during command execution:failed to publish images: error publishing ko://github.com/hellodigit/infra-tools/cert-notifier: writing sbom: PUT https://123456789012.dkr.ecr.us-east-1.amazonaws.com/v2/github.com/some/path/manifests/sha256-a889960bc4557663c9541f6876f3a701708c7f886757a7a407537f8625b7ea9b.sbom: UNSUPPORTED: Invalid parameter at 'ImageManifest' failed to satisfy constraint: 'Invalid JSON syntax'

This looks to be triggered when attempting to upload the SBOM, and happens for a few different builds using v0.12.0. Downgrading to v0.11.2 seems to work fine.

Advice? Any additional information I can grab to help with debugging?

**Comment by imjasonh** [comment]
Between 0.11 and 0.12 we changed the media type that we use when pushing SBOMs: 

https://github.com/ko-build/ko/compare/v0.11.2...v0.12.0#diff-1122f9e78a67758fe70c136cfe87ae6ea4a93ec4dc81a422bfbe57d2321ba868R340

It was `"text/spdx"`, but now it's `"spdx+json"`. I wonder if ECR is not accepting the new media type, or maybe interpreting it differently and failing to parse it somehow.

The JSON we produce should definitely be valid JSON, but maybe the error message is misleading and it's trying to tell us something else.

@michaelb990 @strongjz is this the issue you've been talking about as well?

**Comment by strongjz** [comment]
Yes, because text works fine when I sent it explicitly and not let the file end in .json.

**Comment by frimik** [comment]
Backing down to v0.11.0 for this. What's the solution here? Waiting on ECR to get support or making a special case for the ECR repo SBOM?

**Comment by imjasonh** [comment]
> Backing down to v0.11.2 for this. What's the solution here? Waiting on ECR to get support or making a special case for the ECR repo SBOM?

cc @michaelb990 is there anything we should do? Will ECR support the new `spdx+json` media type?

**Comment by michaelb990** [comment]
Thanks for the ping. This looks similar to the issue that was recently fixed in cosign: https://github.com/sigstore/cosign/pull/2479

The problem is with the media type. `spdx+json` is not a valid media type according to [RFC 6838](https://www.rfc-editor.org/rfc/rfc6838) -- `spdx` should really be a subtype with type of either `text` or `application`. [Both](https://www.iana.org/assignments/media-types/text/spdx) [options](https://www.iana.org/assignments/media-types/application/spdx+json) are how SPDX is registered with IANA.

We are also trying to find some spare cycles to improve the error messages ECR is sending here... I think we could do a lot better :)
